cache.go

```diff
@@ -0,0 +1,200 @@
+package main
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/bitrise-io/go-utils/log"
+	"github.com/bitrise-io/go-utils/pathutil"
+	glob "github.com/ryanuber/go-glob"
+)
+
+// parseIncludeListItem separates path to cache and change indicator path.
+func parseIncludeListItem(item string) (string, string) {
+	// file/or/dir/to/cache -> indicator/file
+	// file/or/dir/to/cache
+	if parts := strings.Split(item, "->"); len(parts) > 1 {
+		return strings.TrimSpace(parts[0]), strings.TrimSpace(parts[1])
+	}
+	return strings.TrimSpace(item), ""
+}
+
+// parseIgnoreListItem separates ignore pattern and if pattern match removes item from cache or not.
+func parseIgnoreListItem(item string) (string, bool) {
+	// path/or/patter/to/exclude
+	// !path/or/patter/to/exclude
+	item = strings.TrimSpace(item)
+	if len(item) > 1 && item[0] == '!' {
+		return strings.TrimSpace(item[1:]), true
+	}
+	return strings.TrimPrefix(item, "!"), false
+}
+
+func parseIncludeList(list []string) map[string]string {
+	indicatorByPath := map[string]string{}
+	for _, item := range list {
+		pth, indicator := parseIncludeListItem(item)
+		if len(pth) == 0 {
+			continue
+		}
+		indicatorByPath[pth] = indicator
+	}
+	return indicatorByPath
+}
+
+func parseIgnoreList(list []string) map[string]bool {
+	ignoreByPath := map[string]bool{}
+	for _, item := range list {
+		pth, ignore := parseIgnoreListItem(item)
+		if len(pth) == 0 {
+			continue
+		}
+		ignoreByPath[pth] = ignore
+	}
+	return ignoreByPath
+}
+
+// expandPath returns every file included in pth (recursively) if it is a dir,
+// if pth is a file it will be returned as an array.
+func expandPath(pth string) ([]string, error) {
+	info, err := os.Lstat(pth)
+	if err != nil {
+		return nil, err
+	}
+
+	if !info.IsDir() {
```

Comment:
a quick example that shows Walk can be executed for file as well and it will iterate on it once:
https://play.golang.org/p/RcE7pAQUxrJ

Reply:
also the builtin function has this: https://golang.org/src/path/filepath/path.go?s=12126:12171#L356